pkg/ctl/delete/nodegroup_test.go

```diff
@@ -0,0 +1,364 @@
+	templateBodyFmt := ""
+	if ngType == api.NodeGroupTypeUnmanaged {
+		templateBodyFmt = UnmanagedNodeGroupTemplateBodyFmt
+	} else if ngType == api.NodeGroupTypeManaged {
+		templateBodyFmt = ManagedNodeGroupTemplateBodyFmt
+	}
+	return fmt.Sprintf(templateBodyFmt, stackName, clusterName, clusterName, stackName, roleARN)
+}
+
+func createAuthConfigMap(resources ...interface{}) (kubernetes.Interface, *v1.ConfigMap) {
+	clientSet := fake.NewSimpleClientset()
+	mapRoles := []string{} // we use the literal declaration as we want "[]"
+	var mapRole string
+	for _, r := range resources {
+		switch r.(type) {
+		case *api.NodeGroup:
+			mapRole = fmt.Sprintf(InstanceRoleFmt, r.(*api.NodeGroup).IAM.InstanceRoleARN)
+		case *api.ManagedNodeGroup:
+			mapRole = fmt.Sprintf(InstanceRoleFmt, r.(*api.ManagedNodeGroup).IAM.InstanceRoleARN)
+		case *api.FargateProfile:
+			mapRole = fmt.Sprintf(InstanceRoleFmt, r.(*api.FargateProfile).PodExecutionRoleARN)
+		default:
+			continue
+		}
+		mapRoles = append(mapRoles, mapRole)
+	}
+	acmData := map[string]string{
+		"mapRoles": strings.Join(mapRoles, "\n"),
+	}
+	acm := &v1.ConfigMap{
+		ObjectMeta: authconfigmap.ObjectMeta(),
+		Data:       acmData,
+	}
+	acm.UID = "12345" // required to set updates and is not provided to us by the fake clientset
+	acm, _ = clientSet.CoreV1().ConfigMaps(authconfigmap.ObjectNamespace).Create(acm)
+	return clientSet, acm
+}
+
+var _ = Describe("removeARN()", func() {
+
+	p := mockprovider.NewMockProvider()
+
+	mockCFN := p.MockCloudFormation()
+
+	mockCFN.On("GetTemplate",
+		mock.MatchedBy(func(input *cloudformation.GetTemplateInput) bool {
+			return input.StackName != nil && *input.StackName == "ng1"
+		})).
+		Return(&cloudformation.GetTemplateOutput{
+			TemplateBody: aws.String(createTemplateBody(api.NodeGroupTypeUnmanaged, "ng1", "my-cluster", ng1.IAM.InstanceRoleARN)),
+		}, nil)
+
+	mockCFN.On("GetTemplate",
+		mock.MatchedBy(func(input *cloudformation.GetTemplateInput) bool {
+			return input.StackName != nil && *input.StackName == "ng1b"
+		})).
+		Return(&cloudformation.GetTemplateOutput{
+			TemplateBody: aws.String(createTemplateBody(api.NodeGroupTypeUnmanaged, "ng1b", "my-cluster", ng1b.IAM.InstanceRoleARN)),
+		}, nil)
+
+	mockCFN.On("GetTemplate",
+		mock.MatchedBy(func(input *cloudformation.GetTemplateInput) bool {
+			return input.StackName != nil && *input.StackName == "ng2"
+		})).
+		Return(&cloudformation.GetTemplateOutput{
+			TemplateBody: aws.String(createTemplateBody(api.NodeGroupTypeUnmanaged, "ng2", "my-cluster", ng2.IAM.InstanceRoleARN)),
+		}, nil)
+
+	mockCFN.On("GetTemplate",
+		mock.MatchedBy(func(input *cloudformation.GetTemplateInput) bool {
+			return input.StackName != nil && *input.StackName == "mng1"
+		})).
+		Return(&cloudformation.GetTemplateOutput{
+			TemplateBody: aws.String(createTemplateBody(api.NodeGroupTypeManaged, "mng1", "my-cluster", ng1.IAM.InstanceRoleARN)),
+		}, nil)
```

Comment:
Ideally, a new mock provider is created for each test case, defining only the mock functions expected in that test. Then if an unexpected call is made, the mock library throws an error and lets us catch unexpected calls.

Reply:
I moved it to a function and call it in each test now